src/vs/platform/dialogs/common/dialogs.ts

```diff
@@ -18,100 +18,166 @@ import { isLinux, isMacintosh, isWindows } from 'vs/base/common/platform';
+	readonly inputArgs?: IInputDialogArgs;
+	readonly promptArgs?: IPromptDialogArgs;
 }
 
-export type DialogType = 'none' | 'info' | 'error' | 'question' | 'warning';
+export interface IBaseDialogOptions {
+	readonly type?: Severity | DialogType;
 
-export interface ICheckbox {
-	readonly label: string;
-	readonly checked?: boolean;
+	readonly title?: string;
+	readonly message: string;
+	readonly detail?: string;
+
+	readonly checkbox?: ICheckbox;
+
+	/**
+	 * Allows to enforce use of custom dialog even in native environments.
+	 */
+	readonly custom?: boolean | ICustomDialogOptions;
 }
 
 export interface IConfirmDialogArgs {
 	readonly confirmation: IConfirmation;
 }
 
-export interface IShowDialogArgs {
-	readonly severity: Severity;
-	readonly message: string;
-	readonly buttons?: string[];
-	readonly options?: IDialogOptions;
+export interface IConfirmation extends IBaseDialogOptions {
+
+	/**
+	 * If not provided, defaults to `Yes`.
+	 */
+	readonly primaryButton?: string;
+
+	/**
+	 * If not provided, defaults to `Cancel`.
+	 */
+	readonly cancelButton?: string;
 }
 
-export interface IInputDialogArgs extends IShowDialogArgs {
-	readonly buttons: string[];
-	readonly inputs: IInput[];
+export interface IConfirmationResult extends ICheckboxResult {
+
+	/**
+	 * Will be true if the dialog was confirmed with the primary button pressed.
+	 */
+	readonly confirmed: boolean;
 }
 
-export interface IDialog {
-	readonly confirmArgs?: IConfirmDialogArgs;
-	readonly showArgs?: IShowDialogArgs;
-	readonly inputArgs?: IInputDialogArgs;
+export interface IInputDialogArgs {
+	readonly input: IInput;
 }
 
-export type IDialogResult = IConfirmationResult | IInputResult | IShowResult;
+export interface IInput extends IConfirmation {
+	readonly inputs: IInputElement[];
+}
 
-export interface IConfirmation {
-	readonly title?: string;
-	readonly type?: DialogType;
-	readonly message: string;
-	readonly detail?: string;
-	readonly primaryButton?: string;
-	readonly secondaryButton?: string;
-	readonly checkbox?: ICheckbox;
+export interface IInputElement {
+	readonly type?: 'text' | 'password';
+	readonly value?: string;
+	readonly placeholder?: string;
 }
 
-export interface IConfirmationResult {
+export interface IInputResult extends IConfirmationResult {
 
 	/**
-	 * Will be true if the dialog was confirmed with the primary button
-	 * pressed.
+	 * Values for the input fields as provided by the user or `undefined` if none.
 	 */
-	readonly confirmed: boolean;
+	readonly values?: string[];
+}
+
+export interface IPromptDialogArgs {
+	readonly prompt: IPrompt<unknown>;
+}
+
+export interface IPromptBaseButton<T> {
 
 	/**
-	 * This will only be defined if the confirmation was created
-	 * with the checkbox option defined.
+	 * @returns the result of the prompt button will be returned
+	 * as result from the `prompt()` call.
 	 */
-	readonly checkboxChecked?: boolean;
+	run(checkbox: ICheckboxResult): T | Promise<T>;
```

Comment:
Nit: I find "run" a slightly odd name for what is essentially a callback. A button cannot "run". Maybe this should be "click" or "onClick"?

Reply:
This is a standard pattern we have for all actions:

https://github.com/microsoft/vscode/blob/19a5794679636e28c36e40bef33838a72cf86f15/src/vs/base/common/actions.ts#L35